recipes-extended/xen/files/memory-scrub-on-shutdown.patch

```diff
@@ -0,0 +1,82 @@
+################################################################################
+SHORT DESCRIPTION:
+################################################################################
+Force page scrub after domain shutdown, not during idle loop.
+
+################################################################################
+LONG DESCRIPTION:
+################################################################################
+In recent Xen versions, page scrubbing has been reworked to occur asynchronously
+during an idle loop. Therefore, it is possible that pages once owned by a guest
+will never have their pages scrubbed if Xen never schedules the loop to run. Xen
+WILL scrub pages before given to a newly booted guest, but during the time the 
+guest is off, it's possible to have unscrubbed pages from previous guests
+hanging about.
+
+So, we reintroduce page scrubbing in free_xenheap_pages and free_domheap_pages,
+so that a guest's pages are scrubbed synchronously when shutdown. We leave
+the idle loop otherwise untouched to reduce the complexity of the patch.
+scrub_free_pages (the logic called from the idle loop) is ultimately called on
+boot when the idle loop is initialized, but after the first run, pages are no
+longer scrubbed from this function.
+
+################################################################################
+CHANGELOG
+################################################################################
+Authors:
+Chris Rogers <rogersc@ainfosec.com>
+
+################################################################################
+REMOVAL
+################################################################################
+Never
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+None, synchronous page scrubbing seems to be a client-side concern, Xen seems
+more concerned with the performance benefit of asynchronous scrubbing.
```

Comment:
This statement needs amending. The concern about scrubbing was raised and discussed on the Xen Community Call; there is awareness of the problem upstream and the committers stated that code to enable selection of the scrubbing behaviour would be welcome.

ie. An option to enable the defensive synchronous scrubbing that OpenXT requires, whilst allowing other Xen users -- particularly those on systems with large amounts of memory that are performing operations such as tests that frequently tear down VMs with lots of memory (think: a database company) -- to retain the background scrub-when-idle option.

Reply:
Ah, I was unaware this was discussed upstream.  Will amend.